File: backend/src/main/java/com/apocalypse/thefall/security/JwtConverter.java
package com.apocalypse.thefall.security;

import org.springframework.core.convert.converter.Converter;
import org.springframework.lang.NonNull;
import org.springframework.security.authentication.AbstractAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Convertit un JWT en un token d'authentification avec ses rôles.
 */
@Component
public class JwtConverter implements Converter<Jwt, AbstractAuthenticationToken> {

    private final JwtGrantedAuthoritiesConverter jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();

    @Override
    public AbstractAuthenticationToken convert(@NonNull Jwt jwt) {
        Collection<GrantedAuthority> defaultAuthorities = jwtGrantedAuthoritiesConverter.convert(jwt);
        Collection<GrantedAuthority> realmAuthorities = extractRealmRoles(jwt);

        Set<GrantedAuthority> allAuthorities = new HashSet<>();
        allAuthorities.addAll(defaultAuthorities);
        allAuthorities.addAll(realmAuthorities);

        return new JwtAuthenticationToken(jwt, allAuthorities);
    }

    /**
     * Extrait les rôles du JWT.
     */
    private Collection<GrantedAuthority> extractRealmRoles(Jwt jwt) {
        return Optional.ofNullable(jwt.getClaimAsMap("realm_access"))
                .map(realmAccess -> realmAccess.get("roles"))
                .filter(Collection.class::isInstance)
                .map(roles -> (Collection<?>) roles)
                .orElseGet(Collections::emptySet)
                .stream()
                .map(Object::toString)
                .map(role -> new SimpleGrantedAuthority("ROLE_" + role))
                .collect(Collectors.toSet());
    }
}

File: backend/src/main/java/com/apocalypse/thefall/security/JwtConverter.kt
package com.apocalypse.thefall.security

import org.springframework.core.convert.converter.Converter
import org.springframework.security.authentication.AbstractAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter
import org.springframework.stereotype.Component

@Component
class JwtConverter : Converter<Jwt, AbstractAuthenticationToken> {

    private val jwtGrantedAuthoritiesConverter = JwtGrantedAuthoritiesConverter()

    override fun convert(jwt: Jwt): AbstractAuthenticationToken {
        val defaultAuthorities: Collection<GrantedAuthority> =
            jwtGrantedAuthoritiesConverter.convert(jwt) ?: emptyList()

        val realmAuthorities = extractRealmRoles(jwt)

        val allAuthorities = (defaultAuthorities + realmAuthorities).toSet()
        return JwtAuthenticationToken(jwt, allAuthorities)
    }

    /**
     * Extracts the roles from the `realm_access.roles` claim and prefixes them with `ROLE_`.
     */
    private fun extractRealmRoles(jwt: Jwt): Collection<GrantedAuthority> {
        val roles = (jwt.getClaimAsMap("realm_access")?.get("roles") as? Collection<*>) ?: emptySet<Any>()
        return roles
            .mapNotNull { it?.toString() }
            .map { role -> SimpleGrantedAuthority("ROLE_$role") }
            .toSet()
    }
}
